libraries/AP_XRCE_Client/AP_XRCE_Client.cpp

```diff
@@ -0,0 +1,118 @@
+#include "AP_XRCE_Client.h"
+
+AP_HAL::Semaphore* AP_XRCE_Client::get_clientsemaphore()
+{
+    return &(AP::ahrs().get_semaphore());
+}
+
+bool AP_XRCE_Client::init()
+{
+    
+    _csem = AP_XRCE_Client::get_clientsemaphore();
+    if (!_csem){
+        hal.console->printf("No semaphore");
+        return false;
+    }
+ 
+    if (this->use_serial) {
+        this->fd=open("/dev/ttyUSB1", O_RDWR | O_NOCTTY | O_NONBLOCK);
+        if(!uxr_init_serial_transport(&this->serial_transport,this->fd,this->relativeSerialAgentAddr,this->relativeSerialClientAddr)) {
+            return false;
+        }
+        uxr_init_session(&this->session, &this->serial_transport.comm, 0xAAAABBBB);
+    }
+    else {
+        if(!uxr_init_udp_transport(&this->udp_transport,UXR_IPv4,ip,port)) {
+            return false;
+        }
+        uxr_init_session(&this->session, &this->udp_transport.comm, 0xAAAABBBB);
+    }
+
+    if (!uxr_create_session(&this->session)) {
+        return false;
+    }
+    
+    if (this->use_serial) {
+        this->reliable_in=uxr_create_input_reliable_stream(&this->session,this->input_reliable_stream,BUFFER_SIZE_SERIAL,STREAM_HISTORY);
+        this->reliable_out=uxr_create_output_reliable_stream(&this->session,this->output_reliable_stream,BUFFER_SIZE_SERIAL,STREAM_HISTORY);
+    }
+    else {
+        this->reliable_in=uxr_create_input_reliable_stream(&this->session,this->input_reliable_stream,BUFFER_SIZE_UDP,STREAM_HISTORY);
+        this->reliable_out=uxr_create_output_reliable_stream(&this->session,this->output_reliable_stream,BUFFER_SIZE_UDP,STREAM_HISTORY);
+    }
+
+    return true;
+}
+bool AP_XRCE_Client::create()
+{
+
+    if(_csem->take(5)){
```

Comment:
taking-with-timeout is almost *always* something you do not want to do.

Also, usually you want to be doing `WITH_SEMAPHORE(_csem)`

Reply:
Done